node/src/proxy_server/mod.rs

```diff
@@ -274,51 +359,49 @@ impl ProxyServer {
                 })
                 .clone()
         };
-        let server_name_opt = return_route_info.server_name_opt.clone();
+
+        let hostname_opt = return_route_info.hostname_opt.clone();
         let response = &msg.payload;
+
         match self.keys_and_addrs.a_to_b(&response.stream_key) {
             Some(socket_addr) => {
-                if let Some(server_name) = server_name_opt {
+                if let Some(server_name) = hostname_opt.clone() {
                     self.subs
                         .as_ref()
                         .expect("Neighborhood unbound in ProxyServer")
                         .update_node_record_metadata
                         .try_send(NodeRecordMetadataMessage {
-                            public_key: exit_public_key.clone(),
+                            public_key: exit_public_key,
                             metadata_change: NRMetadataChange::AddUnreachableHost {
                                 hostname: server_name,
                             },
                         })
                         .expect("Neighborhood is dead");
+                } else {
+                    // TODO: Malefactor ban the exit node because it lied about the DNS failure.
                 }
                 self.report_response_services_consumed(&return_route_info, 0, msg.payload_len);
-
-                self.subs
-                    .as_ref()
-                    .expect("Dispatcher unbound in ProxyServer")
-                    .dispatcher
-                    .try_send(TransmitDataMsg {
-                        endpoint: Endpoint::Socket(socket_addr),
-                        last_data: true,
-                        sequence_number: Some(0), // DNS resolution errors always happen on the first request
-                        data: from_protocol(return_route_info.protocol)
-                            .server_impersonator()
-                            .dns_resolution_failure_response(
-                                &exit_public_key,
-                                return_route_info.server_name_opt.clone(),
-                            ),
-                    })
-                    .expect("Dispatcher is dead");
-                debug!(
-                    self.logger,
-                    "Retiring stream key {}: DnsResolveFailure", &response.stream_key
-                );
-                self.purge_stream_key(&response.stream_key);
+                let retry = match self.pop_dns_failure_retry(&response.stream_key) {
+                    Ok(retry) => retry,
+                    Err(error_msg) => {
+                        panic!("{}", error_msg);
```

Comment:
It's quite similar to another comment, where there's a similar panic. In both these cases it can't be triggered by something outside the network. It is part of the first case you mentioned - _corrupted from inside_.

Reply:
Completed changing `panic!` to `error!`